### PR TITLE
Migrate build system from Makefiles to CMake.

### PR DESCRIPTION
**Pull request by pope**
While this does increase the complexity of the build system, it does unify the build process for all of the environments. It also handles the MozJPEG dependency so that other developers don't have to worry about building it and installing it onto their system. Then there are all of the benefits of using CMake - using other build tools, debug and release build types, and all of the compilation command handling.

I'm not sure if this is event wanted, but I did it and thought I would send the changes back to you to see if it's something you were comfortable with.

### CODE REVIEW

**Comment by lfos** [comment]
I am not generally opposed to moving to another build system. However, there have to be better arguments. If we are replacing our current (very basic and simple) Makefile by something that is more complex and harder to maintain, there need to be convincing benefits. And "other build tools, debug and release build types, and all of the compilation command handling" is nothing that I ever missed so far...

**Comment by pope** [comment]
So the benefits you listed are secondary benefits. I think the main win, and perhaps where most of the complexity comes from, is that it handles MozJPEG consistently across all of the platforms. This downloads the dependency from the git repo, builds it, and links everything to the custom build. I think this is best demonstrated with the `.travis.yml` file - linux users don't need to install MozJPEG to the system just to build this project any more. And it's fewer commands to build everything and get going.

So in summary, the main benefit is that it reduces the amount of documentation to other developers, since more steps are handled by the build process.

**Comment by lfos** [comment]
I wonder whether something as simple as

```diff
diff --git a/Makefile b/Makefile
index d1a2580..1598fe0 100644
--- a/Makefile
+++ b/Makefile
@@ -3,17 +3,16 @@ CFLAGS += -std=c99 -Wall -O3
 LDFLAGS += -lm
 MAKE ?= make
 PREFIX ?= /usr/local
+MOZJPEG_PREFIX ?= mozjpeg
 
 UNAME_S := $(shell uname -s)
 
 ifeq ($(UNAME_S),Linux)
        # Linux (e.g. Ubuntu)
-       MOZJPEG_PREFIX ?= /opt/mozjpeg
        CFLAGS += -I$(MOZJPEG_PREFIX)/include
        LIBJPEG = $(MOZJPEG_PREFIX)/lib/libjpeg.a
 else ifeq ($(UNAME_S),Darwin)
        # Mac OS X
-       MOZJPEG_PREFIX ?= /usr/local/opt/mozjpeg
        LIBJPEG = $(MOZJPEG_PREFIX)/lib/libjpeg.a
        CFLAGS += -I$(MOZJPEG_PREFIX)/include
 else ifeq ($(UNAME_S),FreeBSD)
@@ -59,4 +58,9 @@ install: all
 clean:
        rm -rf jpeg-recompress jpeg-compare jpeg-hash test/test src/*.o src/iqa/build
 
-.PHONY: test install clean
+mozjpeg:
+       test -d mozjpeg || git clone https://github.com/mozilla/mozjpeg.git
+       git -C mozjpeg fetch && git -C mozjpeg checkout v3.3.1
+       ( cd mozjpeg && ./configure && $(MAKE) )
+
+.PHONY: test install clean mozjpeg
```

would achieve pretty much the same thing without all the CMake overhead...

**Comment by pope** [comment]
First, the commands are slightly wrong for building and you haven't set up the dependencies correctly. Also, it doesn't look efficient as you're re-configuring everytime you call mozjpeg. Or do you expect users to type `make mozjpeg` first, and then `make all`?

You need to update `LIBJPEG` as that's in `$(MOZJPEG_PREFIX)/.libs`.

Lastly, this wouldn't work on Windows. Windows builds use CMake. So the output of the lib files would be different. Also, your cleanup files would be different.

It's those little edge cases that I ran into which is usually when I find myself using CMake - which is more verbose but is built to handle all of this stuff, and has those extra benefits I talked about like IDE support and, build types, out-of-source builds, and whatnot.

If you're interested in the general approach of the Makefile handling the MozJPEG dependency, I could create another branch to code that up for a good compare and contrast. Thoughts?

**Comment by pope** [comment]
Here is the Makefile version that pulls in mozjpeg - https://github.com/pope/jpeg-archive/tree/beefy-make.

```diff
diff --git a/Makefile b/Makefile
index d1a2580..74fd758 100644
--- a/Makefile
+++ b/Makefile
@@ -2,51 +2,68 @@ CC ?= gcc
 CFLAGS += -std=c99 -Wall -O3
 LDFLAGS += -lm
 MAKE ?= make
+AUTORECONF ?= autoreconf
+CMAKE ?= cmake
 PREFIX ?= /usr/local
 
 UNAME_S := $(shell uname -s)
-
 ifeq ($(UNAME_S),Linux)
-	# Linux (e.g. Ubuntu)
-	MOZJPEG_PREFIX ?= /opt/mozjpeg
-	CFLAGS += -I$(MOZJPEG_PREFIX)/include
-	LIBJPEG = $(MOZJPEG_PREFIX)/lib/libjpeg.a
+	OS = Unixy
 else ifeq ($(UNAME_S),Darwin)
-	# Mac OS X
-	MOZJPEG_PREFIX ?= /usr/local/opt/mozjpeg
-	LIBJPEG = $(MOZJPEG_PREFIX)/lib/libjpeg.a
-	CFLAGS += -I$(MOZJPEG_PREFIX)/include
+	OS = Unixy
 else ifeq ($(UNAME_S),FreeBSD)
-	# FreeBSD
-	LIBJPEG = $(PREFIX)/lib/mozjpeg/libjpeg.so
-	CFLAGS += -I$(PREFIX)/include/mozjpeg
+	OS = Unixy
+else
+	OS = Windows
+endif
+
+ifeq ($(OS),Unixy)
+	LIBJPEG = src/mozjpeg/.libs/libjpeg.a
 else
-	# Windows
-	LIBJPEG = ../mozjpeg/libjpeg.a
-	CFLAGS += -I../mozjpeg
+	LIBJPEG = src/mozjpeg/libjpeg.a
 endif
+CFLAGS += -Isrc/mozjpeg
 
 LIBIQA=src/iqa/build/release/libiqa.a
 
 all: jpeg-recompress jpeg-compare jpeg-hash
 
+src/mozjpeg:
+	git clone -b v3.3.1 --single-branch https://github.com/mozilla/mozjpeg.git $@
+ifeq ($(OS),Unixy)
+	cd $@ && \
+		$(AUTORECONF) -fiv && \
+		./configure --with-jpeg8
+else
+	cd $@ && \
+		$(CMAKE) -G "MSYS Makefiles" \
+			-DCMAKE_C_COMPILER=$(CC) \
+			-DCMAKE_MAKE_PROGRAM=$(MAKE) \
+			-DCMAKE_BUILD_TYPE=RELWITHDEBINFO \
+			-DWITH_JPEG8=1 .
+endif
+
+$(LIBJPEG): src/mozjpeg
+	cd $< && $(MAKE)
+	touch $@
+
 $(LIBIQA):
 	cd src/iqa; RELEASE=1 $(MAKE)
 
-jpeg-recompress: jpeg-recompress.c src/util.o src/edit.o src/smallfry.o src/commander.o $(LIBIQA)
-	$(CC) $(CFLAGS) -o $@ $^ $(LIBJPEG) $(LDFLAGS)
+jpeg-recompress: jpeg-recompress.c src/util.o src/edit.o src/smallfry.o src/commander.o $(LIBIQA) $(LIBJPEG)
+	$(CC) $(CFLAGS) -o $@ $^ $(LDFLAGS)
 
-jpeg-compare: jpeg-compare.c src/util.o src/hash.o src/edit.o src/commander.o  src/smallfry.o $(LIBIQA)
-	$(CC) $(CFLAGS) -o $@ $^ $(LIBJPEG) $(LDFLAGS)
+jpeg-compare: jpeg-compare.c src/util.o src/hash.o src/edit.o src/commander.o src/smallfry.o $(LIBIQA) $(LIBJPEG)
+	$(CC) $(CFLAGS) -o $@ $^ $(LDFLAGS)
 
-jpeg-hash: jpeg-hash.c src/util.o src/hash.o src/commander.o
-	$(CC) $(CFLAGS) -o $@ $^ $(LIBJPEG) $(LDFLAGS)
+jpeg-hash: jpeg-hash.c src/util.o src/hash.o src/commander.o $(LIBJPEG)
+	$(CC) $(CFLAGS) -o $@ $^ $(LDFLAGS)
 
-%.o: %.c %.h
+%.o: %.c %.h src/mozjpeg
 	$(CC) $(CFLAGS) -c -o $@ $<
 
-test: test/test.c src/util.o src/edit.o src/hash.o
-	$(CC) $(CFLAGS) -o test/$@ $^ $(LIBJPEG) $(LDFLAGS)
+test: test/test.c src/util.o src/edit.o src/hash.o $(LIBJPEG)
+	$(CC) $(CFLAGS) -o test/$@ $^ $(LDFLAGS)
 	./test/$@
 
 install: all
@@ -57,6 +74,13 @@ install: all
 	cp jpeg-hash $(PREFIX)/bin/
 
 clean:
-	rm -rf jpeg-recompress jpeg-compare jpeg-hash test/test src/*.o src/iqa/build
+	rm -rf \
+		jpeg-recompress \
+		jpeg-compare \
+		jpeg-hash \
+		test/test \
+		src/*.o \
+		src/iqa/build \
+		src/mozjpeg
 
 .PHONY: test install clean
```

If that's more to your liking, I'm cool at nuking this and just merging the Makefile only solution.

**Comment by lfos** [comment]
Yes, I like this better. However, I would like to keep the possibility to build against the system mozjpeg (preferably even by default). That is why I suggested to add a `.PHONY` mozjpeg target for users to pull in mozjpeg if required. If you have a better idea, I'd be glad to hear about it!

**Comment by pope** [comment]
So I updated my branch with my latest idea to conditionally build mozjpeg. I use two include files - one to try to find mozjpeg and the other to provide the build rules for mozjpeg if a system copy wasn't found.

I broke these out for readability. I didn't want the main Makefile to be bogged down with mozjpeg rules - the focus should be on the source files.

https://github.com/pope/jpeg-archive/blob/beefy-make/Makefile

**Comment by lfos** [comment]
I will close this for now and we be glad if you could create another PR for that new branch, so that I can review it properly! Thanks!
